validator/src/main/java/com/amazon/aoc/helpers/SortUtils.java

```diff
@@ -0,0 +1,36 @@
+package com.amazon.aoc.helpers;
+
+import com.amazon.aoc.models.xray.Entity;
+
+import java.util.List;
+
+public final class SortUtils {
+
+  /**
+   * Given a list of entities, which are X-Ray segments or subsegments, recursively sort each of
+   * their children subsegments by start time, then sort the given list itself by start time.
+   *
+   * @param entities - list of X-Ray entities to sort recursively. Modified in place.
+   */
+  public static void recursiveEntitySort(List<Entity> entities) {
+    if (entities == null || entities.size() == 0) {
+      return;
+    }
+
+    for (Entity entity : entities) {
+      if (entity.getSubsegments() != null && entity.getSubsegments().size() > 1) {
```

Comment:
I think it's a good idea if we keep recursion depth limit. Major reason being since this is validation tests we can absolutely control the trace data. Also, we wouldn't expect data more than certain depth and if we do program should not go in the loop. I don't see any point of doing recursion after some depth especially in testing scenario. breaking condition looks fine but i strongly suggest we should also break the recursion by some level of depth (maybe 6?).

Reply:
That's a good callout, I can add a limit to prevent infinite looping if, say, a subsegment had itself as a child (which wouldn't be possible in practice but still good to check for). This also made me realize another bug with the stop condition that it shouldn't just ignore if an entity only has a single child, because that single child could itself have many children that need sorting. So I'll address that too.